keps/sig-auth/5681-conditional-authorization/README.md

```diff
@@ -0,0 +1,1425 @@
+
+1. In the authorization stage, the authorizer **partially evaluates** the
+   conditional policies such that the old and new objects and request options
+   are marked unknown (and thus un-evaluatable). If the policy is satisfied *in
+   every other regard* than as a function on the unknown data, this condition on
+   the unknown data is returned by the authorizer to the API server.
+2. If there is a chance of the request becoming authorized, the API server
+   proceeds with the request, and enforces the conditions (as a validating
+   admission step) by either evaluating the conditions in-process (if
+   supported), or by calling the authorizer to evaluate the value of the
+   conditions, now given all possible data.
+
+Notably, this design achieves its goal of unified authorization expressions
+across authorization and admission, without the breaking the reasons why
+authorizers do not have direct access to the request body in the authorization
+stage:
+
+1. Conditional Authorization is only supported for certain requests, namely
+   whenever admission is invoked (verbs create, update, patch, delete).
+1. Any request that cannot become authorized, regardless of the value of the
+   resource data, is rejected already at the authorization stage, thanks to
+   partial evaluation.
+1. The conditions are part of the returned authorization decision, and partial
+   evaluation is a deterministic function, i.e. the same output (which includes
+   the conditions) is guaranteed for the same inputs (metadata and policy store
+   content).
+1. The API server enforces the conditions in the validating admission stage,
+   where access to the objects is available with the correct consistency
+   guarantees.
+
+The abstract Authorizer function definition is augmented from
+`a: Metadata x PolicyStore → Decision` today to
+`a: Metadata x PolicyStore → Decision x ConditionSet`. Concretely, the
+`Decision` enum of the API server gets two new variants: `ConditionalAllow` and
+`ConditionalDeny`. The `authorizer.Authorizer` Go interface’s function signature
+naturally cannot be changed due to backwards compatibility requirements, but
+instead are two dedicated functions used for invoking an authorizer with
+potentially a conditional response, and returning a conditional response from
+the authorizer:
+
+```go
+package authorizer
+
+// Use instead of authorizer.Authorize when the caller wants to use potentially any conditions returned.
+func AuthorizeWithConditionalSupport(ctx context.Context, attrs Attributes, authorizer Authorizer) (Decision, string, ConditionsEnforcer, error)
+
+// A conditional authorizer should return this in its Authorize method, whenever it wants to return a conditional response.
+func NewConditionalDecision(ctx context.Context, self ConditionalAuthorizer, conditions *ConditionSet) (Decision, string, error)
+```
+
+TODO(luxas): Change this to propose another Authorizer interface instead, with upcasting.
```

Comment:
Discussed during the SIG Auth meeting on Nov 19, the meeting was ok with changing the interface such that Authorize returns a `Decision` struct, which contains all data needed, including reason, possible conditionsets, etc.

The downside is of course that importers of `k8s.io/apiserver` will break, the upside is that they will notice this change and be able to adjust as needed, e.g. by taking advantage of this.

Changing the interface is definitely what I wanted to do, I just didn't want to start with that big of a breaking change in the initial prototyping, but I'm glad this was ok for folks, it definitely makes the code more straightforward.